File: src/shared/keepWarm.ts
// Keep-warm / cache-gap diagnostic (P6 + TRDD-VY1IUVUM) — runtime-neutral (host AND webview).
//
// The Anthropic prompt cache expires after a per-session TTL (5 min or 1 hour — see the
// doc-verified matrix in ./cacheTtl). A session whose turns are spaced past ITS TTL re-pays the
// FULL prefix at the cache-WRITE rate (1.25×) on the next call instead of reading it at 0.1× — a
// 12.5× price difference on the dominant token bucket, invisible in per-turn totals because the
// turn "worked". This module measures it per session, from the claude_code.api_request timeline
// entries (the per-call ground truth: exact timestamps + the four disjoint cache buckets).
//
// TTL-AWARENESS (TRDD-VY1IUVUM): the TTL is NOT a global constant. The caller resolves the
// session's TtlRegime (kind × auth × env overrides — ./cacheTtl) and passes it in; a caller with
// no signals gets ASSUMED_TTL_REGIME, and the report SAYS so via ttlSource ('assumed') instead of
// silently guessing. The same 20-min gap is warm on a subscription main session (1h tier) and
// cold on a subagent (5-min tier).
//
// Classification, per consecutive api_request pair (prev → cur), against the EFFECTIVE TTL:
//   • gap < TTL                                → cur is a WARM turn (the cache was still alive).
//   • gap ≥ TTL and cur re-WROTE the prefix    → cur is a COLD turn; its cache_creation tokens are
//     (cacheCreate > cacheRead)                  the measured waste (a warm turn would have READ them).
//   • gap ≥ TTL but no re-write signature      → counted in NEITHER bucket: the TTL passed but no
//                                                penalty was observed (tiny prefix / missing buckets) —
//                                                inventing waste there would be a lie.
// The FIRST api_request of a session follows no gap — its cache write is the unavoidable session
// warm-up, never "waste" — so it is excluded from both buckets by construction.
//
// THE MEASURED FALSIFIER: a cache hit AFTER the assumed expiry contradicts the assumption. If any
// pair shows gap ≥ assumed TTL yet the following call READ a big prefix while writing almost
// nothing (cacheRead ≥ 1k and cacheCreate < cacheRead/4 — the same warm signature the cold-resume
// disarm uses), the entry provably survived that gap. The report then flags ttlContradicted,
// flips ttlSource to 'measured', and classifies against the measured floor: since Anthropic only
// enforces two tiers (5 min / 1 h), a survival past the 5-min tier proves the 1-hour tier — the
// effective TTL snaps to 1h (or to the observed gap itself in the anomalous >1h case) rather than
// to the bare gap, so a 20-min survival doesn't leave 21-min gaps mislabeled cold.
//
// Honest absence: a session with NO api_request entries returns null (rich OTEL logging off, or
// not a Claude Code session) — never zeros presented as measurements.

import type { TimelineEntry } from './summarizerTypes'
import { ASSUMED_TTL_REGIME, TTL_1H_MS, type TtlRegime, type TtlSource } from './cacheTtl'

export interface KeepWarmReport {
  /** Turns that landed inside the effective TTL — the cache was still warm when they hit. */
  warmTurns: number
  /** Turns that followed a ≥TTL gap AND re-wrote the prefix (cacheCreate > cacheRead). */
  coldTurns: number
  /** Σ cache_creation tokens over the cold turns — the prefix re-writes a kept-warm cadence
   *  would have paid at the 0.1× read rate instead of the 1.25× write rate. */
  wastedWriteTokens: number
  /** Largest gap between consecutive api_requests, in minutes (1 decimal). 0 with <2 requests. */
  worstGapMin: number
  /** The TTL the classification ran against, in minutes — the regime's assumption, or the
   *  measured tier when the falsifier contradicted it. */
  ttlAssumedMin: number
  /** Provenance of ttlAssumedMin: doc-matrix | config | measured | assumed — never omitted,
   *  so an absent signal is REPORTED as 'assumed' instead of silently guessed. */
  ttlSource: TtlSource
  /** True when a cache hit after the assumed expiry falsified the regime (details in ttlBasis). */
  ttlContradicted: boolean
  /** The largest gap (minutes, 1 decimal) the cache provably SURVIVED (warm hit after it) while
   *  the assumption said it should have expired. null when nothing contradicted the regime. */
  measuredWarmGapMin: number | null
  /** Provenance sentence for the TTL used (regime basis, or the contradiction evidence). */
  ttlBasis: string
}

/** Warm-hit signature for the falsifier: a big prefix READ with only a small (normal incremental
 *  suffix) write. The 1k floor keeps trivial prefixes from "proving" anything; the /4 share bound
 *  mirrors BodiesActivityTracker.sessionWarmSince (cc < cr × 0.25) so "warm" means the same thing
 *  in the falsifier and in the cold-resume disarm. */
function isWarmHit(cacheRead: number, cacheCreate: number): boolean {
  return cacheRead >= 1_000 && cacheCreate * 4 < cacheRead
}

/**
 * Measure a session's cache keep-warm behaviour from its timeline against its TTL regime.
 * Returns null when the timeline carries no api_request entries with a parseable timestamp
 * (honest absence — see header). Callers without resolved signals omit `regime` and get the
 * assumed 5-min floor, labeled as such.
 */
export function computeKeepWarm(timeline: TimelineEntry[], regime: TtlRegime = ASSUMED_TTL_REGIME): KeepWarmReport | null {
  // Only api_request entries carry the exact per-call timestamps + cache buckets this measurement
  // needs; llm spans can be missing (OTEL down) or lossy. Unparseable timestamps are dropped
  // rather than defaulted — a fabricated ts would fabricate a gap.
  const requests: { ts: number; cacheRead: number; cacheCreate: number }[] = []
  for (const e of timeline) {
    if (e.type !== 'api_request') continue
    const ts = Date.parse(e.timestamp || '')
    if (!Number.isFinite(ts)) continue
    requests.push({ ts, cacheRead: e.cacheReadTokens ?? 0, cacheCreate: e.cacheCreateTokens ?? 0 })
  }
  if (requests.length === 0) return null
  // Timelines are normally chronological, but merged/reparsed sessions can interleave — sort so a
  // spurious negative "gap" can never mis-classify a turn.
  requests.sort((a, b) => a.ts - b.ts)

  // ── Falsifier pass (BEFORE classification, so the contradiction corrects the buckets too):
  // find the largest gap the cache provably survived past the assumed TTL.
  let measuredWarmGapMs = 0
  for (let i = 1; i < requests.length; i++) {
    const gap = requests[i].ts - requests[i - 1].ts
    if (gap >= regime.ttlMs && isWarmHit(requests[i].cacheRead, requests[i].cacheCreate) && gap > measuredWarmGapMs) {
      measuredWarmGapMs = gap
    }
  }
  const contradicted = measuredWarmGapMs > 0
  // Only two tiers exist (doc fact): survival past the assumed tier proves the 1-hour tier.
  // Snapping to the tier (not the bare gap) keeps gaps BETWEEN the observation and the tier edge
  // from being mislabeled cold. A >1h survival is off-matrix (timer-reset semantics or a doc
  // change) — classify against the observation itself rather than inventing a bigger tier.
  const effectiveTtlMs = contradicted ? Math.max(TTL_1H_MS, measuredWarmGapMs + 1) : regime.ttlMs

  let warmTurns = 0
  let coldTurns = 0
  let wastedWriteTokens = 0
  let worstGapMs = 0
  for (let i = 1; i < requests.length; i++) {
    const gap = requests[i].ts - requests[i - 1].ts
    if (gap > worstGapMs) worstGapMs = gap
    if (gap < effectiveTtlMs) {
      warmTurns++
    } else if (requests[i].cacheCreate > requests[i].cacheRead) {
      // The TTL expired AND this call re-wrote more prefix than it read — the cold signature.
      coldTurns++
      wastedWriteTokens += requests[i].cacheCreate
    }
    // gap ≥ TTL without the re-write signature: neither bucket (no observed penalty — header note).
  }
  const toMin = (ms: number): number => Math.round(ms / 60_000 * 10) / 10
  return {
    warmTurns,
    coldTurns,
    wastedWriteTokens,
    worstGapMin: toMin(worstGapMs),
    ttlAssumedMin: Math.round(effectiveTtlMs / 60_000),
    ttlSource: contradicted ? 'measured' : regime.ttlSource,
    ttlContradicted: contradicted,
    measuredWarmGapMin: contradicted ? toMin(measuredWarmGapMs) : null,
    ttlBasis: contradicted
      ? `assumed ${regime.ttlAssumedMin}-min TTL (${regime.ttlSource}) CONTRADICTED: a cache hit landed ` +
        `${toMin(measuredWarmGapMs)}min after the last call — the entry survived, so the measured floor ` +
        `(${Math.round(effectiveTtlMs / 60_000)}min tier) was preferred`
      : regime.ttlBasis,
  }
}

File: rust-core/crates/agentlens-core/src/burn/keep_warm.rs
//! Port of src/shared/keepWarm.ts (P6 + TRDD-VY1IUVUM) — the per-session cache keep-warm
//! measurement over `claude_code.api_request` timeline entries, classified against the
//! session's TtlRegime, with THE MEASURED FALSIFIER: a cache hit AFTER the assumed expiry
//! contradicts the assumption, flips ttlSource to 'measured' and snaps the effective TTL to
//! the 1-hour tier (only two tiers exist), or to the observed gap itself past 1h.
//!
//! Honest absence: a timeline with NO parseable api_request entries returns None — never
//! zeros presented as measurements. The report is emitted as a `Value` mirroring the TS
//! object literal (the P4d design law: no typed-struct mirror of the wire shape).

use serde_json::{Map, Value};

use super::cache_ttl::{TtlRegime, TTL_1H_MS};
use crate::summarize::helpers;

/// Warm-hit signature for the falsifier: a big prefix READ with only a small (normal
/// incremental suffix) write. The 1k floor keeps trivial prefixes from "proving" anything;
/// the /4 share bound mirrors BodiesActivityTracker.sessionWarmSince so "warm" means the same
/// thing in the falsifier and in the cold-resume disarm.
pub fn is_warm_hit(cache_read: f64, cache_create: f64) -> bool {
    cache_read >= 1_000.0 && cache_create * 4.0 < cache_read
}

/// computeKeepWarm — measure a session's cache keep-warm behaviour from its timeline against
/// its TTL regime. None when the timeline carries no api_request entry with a parseable
/// timestamp. Classification per consecutive pair, against the EFFECTIVE TTL:
/// gap < TTL → warm; gap ≥ TTL with cacheCreate > cacheRead → cold (+wasted writes);
/// gap ≥ TTL without the re-write signature → NEITHER bucket (no observed penalty). The first
/// request follows no gap — the unavoidable session warm-up, excluded by construction.
pub fn compute_keep_warm(timeline: &[Value], regime: &TtlRegime) -> Option<Value> {
    // Only api_request entries carry the exact per-call timestamps + cache buckets this needs.
    // Unparseable timestamps are DROPPED, not defaulted — a fabricated ts would fabricate a gap.
    let mut requests: Vec<(f64, f64, f64)> = Vec::new();
    for e in timeline {
        if e.get("type").and_then(Value::as_str) != Some("api_request") {
            continue;
        }
        let ts_str = e.get("timestamp").and_then(Value::as_str).unwrap_or("");
        let Some(ts) = helpers::parse_iso_ms(ts_str) else { continue };
        let bucket = |k: &str| e.get(k).and_then(Value::as_f64).unwrap_or(0.0); // `?? 0`
        requests.push((ts, bucket("cacheReadTokens"), bucket("cacheCreateTokens")));
    }
    if requests.is_empty() {
        return None;
    }
    // Merged/reparsed sessions can interleave — sort so a spurious negative "gap" can never
    // mis-classify a turn (ts values are finite: the parse filter above dropped NaN).
    requests.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(std::cmp::Ordering::Equal));

    // Falsifier pass BEFORE classification, so the contradiction corrects the buckets too.
    let mut measured_warm_gap_ms = 0.0f64;
    for i in 1..requests.len() {
        let gap = requests[i].0 - requests[i - 1].0;
        if gap >= regime.ttl_ms && is_warm_hit(requests[i].1, requests[i].2) && gap > measured_warm_gap_ms {
            measured_warm_gap_ms = gap;
        }
    }
    let contradicted = measured_warm_gap_ms > 0.0;
    // Survival past the assumed tier proves the 1-hour tier (only two tiers exist); a >1h
    // survival is off-matrix — classify against the observation itself.
    let effective_ttl_ms = if contradicted { TTL_1H_MS.max(measured_warm_gap_ms + 1.0) } else { regime.ttl_ms };

    let mut warm_turns = 0.0f64;
    let mut cold_turns = 0.0f64;
    let mut wasted_write_tokens = 0.0f64;
    let mut worst_gap_ms = 0.0f64;
    for i in 1..requests.len() {
        let gap = requests[i].0 - requests[i - 1].0;
        if gap > worst_gap_ms {
            worst_gap_ms = gap;
        }
        if gap < effective_ttl_ms {
            warm_turns += 1.0;
        } else if requests[i].2 > requests[i].1 {
            // The TTL expired AND this call re-wrote more prefix than it read — the cold signature.
            cold_turns += 1.0;
            wasted_write_tokens += requests[i].2;
        }
    }
    // `Math.round(ms / 60_000 * 10) / 10` — JS half-up equals f64::round for non-negative gaps.
    let to_min = |ms: f64| (ms / 60_000.0 * 10.0).round() / 10.0;
    let effective_min = (effective_ttl_ms / 60_000.0).round();
    let mut m = Map::new();
    m.insert("warmTurns".into(), helpers::num(warm_turns));
    m.insert("coldTurns".into(), helpers::num(cold_turns));
    m.insert("wastedWriteTokens".into(), helpers::num(wasted_write_tokens));
    m.insert("worstGapMin".into(), helpers::num(to_min(worst_gap_ms)));
    m.insert("ttlAssumedMin".into(), helpers::num(effective_min));
    m.insert("ttlSource".into(), if contradicted { "measured".into() } else { regime.ttl_source.into() });
    m.insert("ttlContradicted".into(), Value::Bool(contradicted));
    m.insert("measuredWarmGapMin".into(), if contradicted { helpers::num(to_min(measured_warm_gap_ms)) } else { Value::Null });
    m.insert(
        "ttlBasis".into(),
        if contradicted {
            format!(
                "assumed {}-min TTL ({}) CONTRADICTED: a cache hit landed {}min after the last call — the entry survived, so the measured floor ({}min tier) was preferred",
                helpers::fmt_js_num(regime.ttl_assumed_min),
                regime.ttl_source,
                helpers::fmt_js_num(to_min(measured_warm_gap_ms)),
                helpers::fmt_js_num(effective_min)
            )
            .into()
        } else {
            regime.ttl_basis.clone().into()
        },
    );
    Some(Value::Object(m))
}
